Parse/src/main/java/com/parse/ParseQuery.java

```diff
@@ -960,33 +960,21 @@ public ParseQuery(ParseQuery<T> query) {
     return ParseUser.getCurrentUserAsync();
   }
 
-  private void checkIfRunning() {
-    checkIfRunning(false);
-  }
-
-  private void checkIfRunning(boolean grabLock) {
-    synchronized (lock) {
-      if (isRunning) {
-        throw new RuntimeException(
-            "This query has an outstanding network connection. You have to wait until it's done.");
-      } else if (grabLock) {
-        isRunning = true;
-        cts = Task.create();
-      }
-    }
-  }
-
   /**
    * Cancels the current network request (if one is running).
    */
   //TODO (grantland): Deprecate and replace with CancellationTokens
```

Comment:
On top of that , if they wanted to deprecate `query.cancel()`, they probably meant to have the developer get / pass a `CancellationToken` and call `source.cancel()` so the token is notified. See also the comment [here](https://github.com/parse-community/Parse-SDK-Android/blob/master/Parse/src/main/java/com/parse/ParseQuery.java#L1472).

The same comment is left in `ParseFile` e.g. [here](https://github.com/parse-community/Parse-SDK-Android/blob/master/Parse/src/main/java/com/parse/ParseFile.java#L709). We might set up a PR that addresses this in both classes. But the only way I see this possible is overloading methods, like

```java
CancellationTokenSource source = new CancellationTokenSource();
query.findInBackground(callback, /* nullable */ source.getToken());
// At any moment..
source.cancel();
```

It would be useful because `ParseFile` and with this PR `ParseQuery` allow multiple concurrent operations, so passing a token to each lets the developer cancel only one operation. If you think it’s needed I’m going to work on it (separately), with the plan

- overload methods to accept a token
- deprecate `cancel()` and have it fallback to `cancelAll()`
- create `cancelAll()` that cancels all currently running task

So the developer either passes a cancellation token to a single operation or uses `cancelAll` if he doesn’t care.